mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs

```diff
@@ -457,37 +458,120 @@ impl EthCoin {
+            })?;
+        let approve_func = TAKER_SWAP_V2.function(TAKER_PAYMENT_APPROVE)?;
+        let decoded = decode_contract_call(approve_func, taker_payment.unsigned().data())?;
+        let id = match decoded.first() {
+            Some(Token::FixedBytes(bytes)) => bytes,
+            invalid_token => {
+                return MmError::err(FindPaymentSpendError::InvalidData(format!(
+                    "Expected Token::FixedBytes, got {:?}",
+                    invalid_token
+                )))
+            },
+        };
+        // loop to find maker's spendTakerPayment transaction
         loop {
-            let taker_status = self
-                .payment_status_v2(
-                    taker_swap_v2_contract,
-                    decoded[0].clone(), // id from spendTakerPayment
-                    &TAKER_SWAP_V2,
-                    EthPaymentType::TakerPayments,
-                    TAKER_PAYMENT_STATE_INDEX,
-                )
-                .await?;
-            if taker_status == U256::from(TakerPaymentStateV2::MakerSpent as u8) {
-                return Ok(taker_payment.clone());
-            }
             let now = now_sec();
             if now > wait_until {
-                return MmError::err(WaitForPaymentSpendError::Timeout { wait_until, now });
+                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
             }
-            Timer::sleep(10.).await;
+
+            let current_block = match self.current_block().compat().await {
+                Ok(b) => b,
+                Err(e) => {
+                    error!("Error getting block number: {}", e);
+                    Timer::sleep(5.).await;
+                    continue;
+                },
+            };
+
+            // get all logged TakerPaymentSpent events from `from_block` till current block
+            let events = match self
+                .events_from_block(taker_swap_v2_contract, "TakerPaymentSpent", from_block, current_block)
+                .await
+            {
+                Ok(events) => events,
+                Err(e) => {
+                    error!(
+                        "Error getting TakerPaymentSpent events from {} block: {}",
+                        from_block, e
+                    );
+                    Timer::sleep(5.).await;
+                    continue;
+                },
+            };
+
+            if events.is_empty() {
+                info!(
+                    "No events found yet for the block range {} to {}",
+                    from_block, current_block
+                );
+                Timer::sleep(5.).await;
+                continue;
+            }
+
+            // this is how spent event looks like in EtomicSwapTakerV2: event TakerPaymentSpent(bytes32 id, bytes32 secret)
+            let found_event = events.into_iter().find(|event| &event.data.0[..32] == id.as_slice());
```

Comment:
I think the code part to find the event:
```
let current_block = match self.current_block().compat().await {
...
let found_event = events.into_iter().find(|event| &event.data.0[..32] == id.as_slice());
```
..could be moved in a function.
So we could avoid multiple calls to Timer::sleep(5.).await
(and maybe that fn even could be reusable)

Reply:
Could you re review events handling, I removed get current_block function and refactored `tx_hash` usage, feels like there is no need in standalone priv function.

Reply:
Okay,
could you plz take a look how the same waiting for tx code implemented in legacy swaps, for e.g. search_for_swap_tx_spend():
- no Timer::sleep(5.).await
- if self.transaction returns an error the fn exits
- from_block resets to current_block for each loop 
There are 3 implementations of similar code btw.
Maybe all could be made as function/generic

Reply:
The thing is that `search_for_swap_tx_spend` is used inside `search_for_swap_tx_spend_my` and `search_for_swap_tx_spend_other` functions, which are used in recover funds and in taker restart processes. 
It could be after timelock, thats why there is no time sleep and timeout check and error from self.transaction returns  stops loop.

So its not for the actual swap stage when user was notified that maker spent payment and taker started searching Maker's tx to extract secret.

Thats ^ why we have `wait_for`**_htlc_**`tx_spend` and `search_for_swap_tx_spend` functions in legacy. But its still difficult to understand functions purposes just looking at their names without doc comments and studying the code.

> from_block resets to current_block for each loop

I wouldn't move blocks range on HTLC functionality due to this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261#discussion_r1839934253 Its too risky to trust web3 api and blockchain that there will be no tx_hash Nones when taker almost successfully finished their swap without problems on previous stages.

I suggest not to touch `search_for_swap_tx_spend` as it has different logic.

For now removed current_block call from `wait_for_htlc_tx_spend` and reused `events_from_block` function, instead of legacy spend_events function https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/24cb4c3fc5d5572d66548aee65822a95463df481

Reply:
I noticed that the legacy event loop uses an event lookup limit: logs_block_range. Looks like providers impose different limits on log ranges. Shouldn't we apply it here too?
If `find_taker_payment_spend_tx` actually does waiting maybe it still should be named as `wait_for...` (we discussed in DM that find is okay but I did not realise it would be waiting internally, well it's up to you ofc)?

Reply:
> I noticed that the legacy event loop uses an event lookup limit: logs_block_range. 

Note that you are referencing to `search_for_swap_tx_spend` legacy function. 
Other legacy `wait_for_htlc_tx_spend` doesnt use range. I suppose its bcz `wait_for_htlc_tx_spend` is used in swap stage where trader continues their swap without problems, so there are not a lot of new blocks from previous stage.
When trader have to call `search_for_swap_tx_spend` (they need to refund or recreate swap), there could be much more new blocks, which lager than "eth_getLogs" limit

I added block windows in tpu v2 function https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/5f666f3d14a879314bef9f864b06de3256337ac6

and in legacy https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/47f904ceebd27a7d1d917f436c8ec0861579d2c6  well in theory  `wait_for_htlc_tx_spend` could see limits problem in some very active blockchain. But I wasnt there when legacy code was created, so legacy code could be reverted if we decide not to change it.

> If find_taker_payment_spend_tx actually does waiting 

It has `wait_until` field, however its is used for timeout, the actual logic is to research log events to find matched swap id to get tx_hash. and only then we get the transaction. Its not the same as `wait_for_confirmation` (as an example), as here we already know tx_hash and just wait for confirmations.

Reply:
I think we should refactor all/some existing eth event loops and make a generic function (with params) which can be used for all cases.
But I guess this is a task for another PR, for refactoring. I suggest adding a TODO for that.